### ISSUE TITLE
clarifying the canonical representation of language-tagged string

### ISSUE THREAD

**Issue by pchampin** [w3c/rdf-canon#155]
I was surprised to read, in [Appendix B: URDNA 2015](https://www.w3.org/TR/rdf-canon/#urdna2015), that

> The version of the algorithm defined in [A. A Canonical form of N-Quads](https://www.w3.org/TR/rdf-canon/#canonical-quads) clarifies the representation of [language-tagged strings](https://www.w3.org/TR/rdf11-concepts/#dfn-language-tagged-string)

because in fact, language-tagged-strings are not even mentioned in Appendix A, nor anywhere else in the spec.

In particular, it is not clear whether `"foo"@en` and `"foo"@EN` are expected to be represented differently or identically in canonical N-Quads (but a few implementations I have tested represent them both as `"foo"@en`). Of course, this not only impact the final representation of the canonicalized dataset, it also impacts hashing!

For example, the following input:
```n-quads
_:e1 <tag:p> "foo"@EN .
_:e2 <tag:p> "foo"@FR .
```
canonicalizes to
```n-quads
_:c14n0 <tag:p> "foo"@EN .
_:c14n1 <tag:p> "foo"@FR .
```
if the case of language tags is preserved, but to
```n-quads
_:c14n0 <tag:p> "foo"@fr .
_:c14n1 <tag:p> "foo"@en .
```
if language tags are lowercased -- noticed how the order of bnodes has changed!

I think that we need:
- to add the missing clarification about language-tagged strings in Appendix A
- add corresponding test cases (currently, all language tags in the tests are in lowercase, so we have a blind spot)

**Comment by gkellogg** [comment]
The appendix comes directly from [A Canonical form of N-Quads](https://www.w3.org/TR/rdf-n-quads/#canonical-quads) in [RDF 1.2 N-Quads](https://www.w3.org/TR/rdf-n-quads/). Although rdf-canon uses this definition, it is working on RDF 1.1. The status of language tags in RDF 1.1 is unclear, as [Literals](https://www.w3.org/TR/rdf-concepts/#section-Graph-Literal) in [RDF 1.2 Concepts](https://www.w3.org/TR/rdf-concepts) and [RDF 1.1 Concepts](https://www.w3.org/TR/rdf11-concepts/#section-Graph-Literal) that language tags _MAY_ be converted to lower case, but [Literal term equality](https://www.w3.org/TR/rdf11-concepts/#dfn-literal-term-equality) says that they are compared character by character. So, as it stands in RDF 1.1, arguably `"foo"@en-us` and `"foo"@en-US` are two different RDF terms, but express the same value via D-entailment.

There is w3c/rdf-concepts#55 to change the way these terms are compared, which would form a better basis for updating canonicalization in N-Triples, N-Quads, and in rdf-canon to perform such canonicalization, so we should probably wait until that has been settled, even though that will be for RDF 1.2.

An alternative to normalizing to lower case would be to use [BCP47 Canonicalization of Language Tags](https://www.rfc-editor.org/rfc/rfc5646.html#section-4.5), but this might prove challenging.

**Comment by iherman** [comment]
Ouch.

While I agree with @gkellogg that, in essence, this is not our problem 😀, meaning that we inherit the ambiguities of RDF 1.1, I think it would be worthwhile to put a note into the spec drawing attention to this issue and to the fact that RDF 1.2 should settle the problem. If RDF 1.2 is out, and implementation (essentially, N-Quad parsers and serializers) are updated, RDFC 1.0 would automatically work. The note might also include the advice that applications relying on RDFC 1.0 should adopt a systematic choice of handling language tags, e.g., by systematically set them to lower case.

The alternative that @gkellogg referred to is that an RDFC 1.0 would always use, e.g., lower case language tags when hashing, but that might be a bit tedious because it may involve meddling with the underlying RDF framework that the implementation relies on...

**Comment by gkellogg** [comment]
I think the only thing we can do is to direct readers to the issue; given literals that vary only in the case used in language tags, some implementations will normalize them to lower case and others won't. In fact, this sets up an isomorphism issue in RDF 1.1 that I don't see how RDF 1.2 can really resolve, as changing to require all implementations to normalize to lower case would likely face objections, but we'll certainly discuss.

This also creates a canonicalization issue, in that some processors may normalize language tags and others may not (given the MAY clause in RDF Concepts). Normalizing to lower case during canonicalization would potentially mean that the canonical form of a graph is not isomorphic to the original, so I don't think we can simply add language to normalize language tags when serializing. In any case, it would need to be done on input so that the hashes came out the same.

We could add a clause that input datasets MUST use language tags that are normalized to lower case, or, we could require that processors normalize language tags to lower case (making the  MAY a MUST), but probably as @iherman suggests, an informative warning about the issue is about all we can manage.

**Comment by pchampin** [comment]
@ivan @gkellogg generally +1 to what you wrote.

However, shouldn't we remove the confusing text that says :

> The version of the algorithm defined in [A. A Canonical form of N-Quads](https://www.w3.org/TR/rdf-canon/#canonical-quads) clarifies the representation of [language-tagged strings](https://www.w3.org/TR/rdf11-concepts/#dfn-language-tagged-string)

when in fact Appendix A does not mention language tags at all??

**Comment by gkellogg** [comment]
I was hoping to get some clarification from the RDF-star WG on a direction for language tag normalization, but I think summer vacations may get in the way of that.

At this point, I think the best we can do is to warn about the potential for variation if language tags are not normalized to lower case on the input. We can't change the requirements of parsers on handling language tags, so that one N-Quads parser may represent a literal as `"foo"@en-US` and another as `"foo"@en-us`. Depending on the parser, that could result in one quad or two quads, result in different hashes of those quads and datasets and make one dataset which would otherwise be considered to be isomorphic to another result in different hashes and canonical representations.

Regarding the appendix, we should clarify (possibly with a reference to that note) that `LANGTAG` is _not_ normalized.

A stronger solution would be to error on any input dataset which included a language-tagged string containing a `LANGTAG` which did not consist entirely of lower-case characters matching BCP47.

**Comment by gkellogg** [comment]
> However, shouldn't we remove the confusing text that says :
> 
> > The version of the algorithm defined in [A. A Canonical form of N-Quads](https://www.w3.org/TR/rdf-canon/#canonical-quads) clarifies the representation of [language-tagged strings](https://www.w3.org/TR/rdf11-concepts/#dfn-language-tagged-string)
> 
> when in fact Appendix A does not mention language tags at all??

It doesn't really need any normative language for language-tagged strings, other than an addition to the second point that literals with neither `xsd:string` or `rdf:langString` may not use the datatype IRI; the fact that the `LANGTAG` component is part of the serialization is implicit from the grammar. What needs to be added, based on our direction, is that the value matching `LANGTAG` MUST NOT be modified or normalized. This change needs to go into RDF 1.2 N-Quads and N-Triples as well.

**Comment by gkellogg** [comment]
The related issues in N-Quads (w3c/rdf-n-quads#45) and N-Triples (w3c/rdf-n-triples#41) were closed as the issue is one of potential normalization on input, rather than output. This narrows the issue for rdf-canon to discussing norms on the input datasets.

We may want to add something to Security Considerations about false negatives stemming from different representations of isomorphic datasets which can yield different canonical forms.

**Comment by gkellogg** [comment]
Fixed by #156.
